waiter/src/waiter/metrics_sync.clj

```diff
@@ -104,10 +107,58 @@
         (log/error "not registering request as it is missing request id")
         router-metrics-state))))
 
+(defn- clean-service-id->instance-id->metric
+  "Remove services from outer map that are not tracked by this router. Remove instances in the instance-id->metric map
+  that are not tracked by this router. This is done to prevent memory leaks in the service-id->instance-id->metric map"
+  [query-state-fn service-id->instance-id->metric]
+  (let [{:keys [all-available-service-ids service-id->healthy-instances service-id->unhealthy-instances]} (query-state-fn)]
+    (->> (select-keys service-id->instance-id->metric all-available-service-ids)
+         seq
+         (map
+           (fn [[service-id instance-id->metric]]
+             [service-id
+              (select-keys
+                instance-id->metric
+                (map
+                  :id
+                  (concat
+                    (get service-id->healthy-instances service-id)
+                    (get service-id->unhealthy-instances service-id))))]))
+         (into {}))))
+
+(defn- merge-instance-id->metric-maps
+  "Merges two instance-id->metric maps iterating through the instance-ids in both maps and checking which map has the
+  latest metric based on 'updated-at'. If the metric does not exist in one of the maps, then defaults to the non nil
+  metric."
+  [instance-id->metric-1 instance-id->metric-2]
+  (merge-with (fn [metric-1 metric-2]
+                (let [updated-at-1 (get metric-1 "updated-at")
+                      updated-at-2 (get metric-2 "updated-at")]
+                  ; We can assume that 'updated-at' is an ISO-8601 timestamp because it was already validated
+                  (if (pos? (compare updated-at-2 updated-at-1))
+                    metric-2
+                    metric-1)))
+              instance-id->metric-1
+              instance-id->metric-2))
+
+(defn- merge-service-id->instance-id->metric-maps
```

Comment:
Nitpick: Can we clean after the merge, feels like we can avoid some computation there?